### PR TITLE
Decoders improvments

### PR DESCRIPTION
**Pull request by jrouaix**
Hi @puffyCid,

This time I started some work on the `decoders` module.

It's a draft for now, I left a lot of todos in the `dns` file.

I think there are some subtleties I don't get. 
Something I have the intuition that the same data is parsed multiple times, perhaps with multiple formats, it's very weird.

Also, I have some troubles with the error handling.
Replacing `Ok` results with some default error String will bite in the end. 
We cannot know from the result of a function if the parsing was right. 
It's probably not the right moment for a refacto on this topic, but in the end, I think we could tend to a design where `Result<_, _>`s are first class citizens as return values or even fields in the parsed structures, so the client code can choose what behavior to adopt.

what do you think ?

### CODE REVIEW

**Comment by puffyCid** [comment]
@jrouaix for the dns.rs` file `get_dns_flags` is parsing bits not bytes. We nom 2 bytes first and then pass that to `get_dns_flags` which then pareses 13 bits

It is a bit weird but there is a small comment mentioning it:
```
    // Have to work with bits instead of bytes for the DNS flags
    let ((flag_data, offset), query_flag): ((&[u8], usize), u8) = bits::complete::take(size_of::<u8>())((data, 0))?;
```
https://en.wikipedia.org/wiki/Domain_Name_System#DNS_message_format

I do not think we are parsing the same data multiple times. 
```
let (dns_data, id_data) = take(size_of::<u16>())(data)?;
let (dns_data, flag_data) = take(size_of::<u16>())(dns_data)?;

let (_, id) = be_u16(id_data)?;

let message_result = get_dns_flags(flag_data);
let message = match message_result 
...
 let message_result = parse_counts(dns_data);
```

For above `dns_data` is simply the remaining bytes.
Ex: 
```
data = 10 bytes
let (dns_data, id_data) = take(size_of::<u16>())(data)?;

id_data = 2 bytes
dns_data = 8 bytes

let (dns_data, flag_data) = take(size_of::<u16>())(dns_data)?;

flag_data = 2 bytes
dns_data = 6 bytes

etc
```

For return string vs Result I am open to changes. The main thing i think we should be aware is IMO I think iteration should not stop even we get an error. Ex: If the library is looping/decoding through 200 entries and it fails to decode something, the library should log an error and include something (ex: partial string, original data, etc) vs return error/breaking the loop.

I am open to changing what the decoders should do if it encounters an error on whether to use Result vs String, but as long as overall iteration continues and we do not skip any entries if an error is encountered.

Let me know if this makes sense or helps
thanks

**Comment by jrouaix** [comment]
HOOO !

Got it, you use `size_of::<u8>()` as expression for `1` !
But `size_of::<u8>()` is meant to be understood as `bytes` => u8 is ` byte long. not 1 bit.
Thats why I was kinda sure you were consuming bytes !

It's change that to `const` values if you don't mind.

**Comment by jrouaix** [comment]
@puffyCid you can have a look ?

I spent some more time in the `dns.rs` file.

Now `decoder.rs` expect `Result<>`'s from the sub decoders

I also tried to avoid or delay allocation (strings re-allocations in loops), in the long run it will get us some perf improvements. 

I lost some precision about errors location but I find the overall design i little nicer.

Tell me what you think, perhaps it's mergable now. As you wish.

**Comment by puffyCid** [comment]
> @puffyCid you can have a look ?
> 
> I spent some more time in the `dns.rs` file.
> 
> Now `decoder.rs` expect `Result<>`'s from the sub decoders
> 
> I also tried to avoid or delay allocation (strings re-allocations in loops), in the long run it will get us some perf improvements.
> 
> I lost some precision about errors location but I find the overall design i little nicer.
> 
> Tell me what you think, perhaps it's mergable now. As you wish.

PR looks good. Thanks. Left just one minor comment.

I recently merged #38 to fix an infinite loop bug.
If u could resolve the conflicts, and mark the PR as non-draft. I fine with merging

**Comment by jrouaix** [comment]
@puffyCid I encourage you to squash merge this PR, commits are a little messy

File: src/decoders/dns.rs [hunk 1]
```diff
@@ -147,245 +136,250 @@ fn get_dns_flags(data: &[u8]) -> nom::IResult<(&[u8], usize), String> {
...
+        message.push_str(&alpn_name);
+        message.push(',')
     }
-    Ok((data, message))
+    Ok((input, message))
 }
 
 /// Parse the IPs
-fn parse_svcb_ip(data: &[u8]) -> nom::IResult<&[u8], String> {
+// fn parse_svcb_ip(data: &[u8]) -> nom::IResult<&[u8], String> {
```
Comment (puffyCid) [inline src/decoders/dns.rs:230]: Minor nitpick, but i would prefer for now if this PR did not make changes to the function `parse_svcb_ip` for now

Both #38 and this PR make changes to dns.rs. This PR makes alot more. 
Once this PR gets merged, i am totally fine if you want open another PR to make some fixes/changes to `parse_svcb_ip`

#38 fixes an infinite loop bug, i just want to make sure these changes do not revert that fix. Opening a smaller PR will make the review alot easier IMO.

File: src/decoders/decoder.rs [hunk 2]
```diff
@@ -23,108 +23,124 @@ use crate::{
...
-        message_value = dns_getaddrinfo_opts(&message_values[index].message_strings);
+        Ok(dns_getaddrinfo_opts(&message_values[index].message_strings))
+    } else {
+        Ok(String::new())
+    };
+
+    match message_value {
+        Ok(value) => value,
+        Err(e) => {
+            log::error!("[macos-unifiedlogs] Failed to parse DNS header counts. Error: {e:?}");
```
[the remark this answers is not in the document]
Comment (jrouaix) [reply src/decoders/decoder.rs:140]: Yes ! copy/pasta error.

File: src/decoders/dns.rs (hunk 1 above)
Comment (jrouaix) [reply src/decoders/dns.rs:230]: Ok, added a test to ensure the current behavior of this function and creating a new PR with only my newest version

File: src/decoders/network.rs [hunk 3]
```diff
@@ -150,21 +150,17 @@ fn get_sockaddr_data(data: &[u8]) -> nom::IResult<&[u8], String> {
...
 /// Get the IPv6 data
-pub(crate) fn get_ip_six(data: &[u8]) -> nom::IResult<&[u8], String> {
-    let (ip_data, ip_addr_data) = take(size_of::<u128>())(data)?;
-    let (_, ip) = be_u128(ip_addr_data)?;
+pub(crate) fn get_ip_six(input: &[u8]) -> nom::IResult<&[u8], String> {
+    let (input, ip) = be_u128(input)?;
     let ip = Ipv6Addr::from(ip);
-
-    Ok((ip_data, ip.to_string()))
+    Ok((input, ip.to_string()))
```
Comment (dgmcdona) [inline src/decoders/network.rs:163]: This function, as well as `get_ip_four`, can be shortened like this:

```rust
use nom::combinator::map;

map(be_u128, |val| Ipv6Addr::from(val).to_string())(input)
```

Comment (jrouaix) [reply src/decoders/network.rs:163]: also indeed

**Comment by puffyCid** [comment]
merged thanks!

**Pull request merged**